flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ExternalSchemaTranslator.java

```diff
@@ -154,6 +154,26 @@ public static InputResult fromExternal(
             @Nullable Schema declaredSchema) {
         final DataType inputDataType =
                 TypeInfoDataTypeConverter.toDataType(dataTypeFactory, inputTypeInfo);
+        return fromExternal(dataTypeFactory, inputDataType, declaredSchema);
+    }
+
+    /**
+     * Converts the given {@link DataType} and an optional declared {@link Schema} (possibly
+     * incomplete) into the final {@link InputResult}.
+     *
+     * <p>This method serves three types of use cases:
+     *
+     * <ul>
+     *   <li>1. Derive physical columns from the input type information.
+     *   <li>2. Derive physical columns but merge them with declared computed columns and other
+     *       schema information.
+     *   <li>3. Derive and enrich physical columns and merge other schema information.
```

Comment:
we should disable the third use case for `Table#executeInsert` via a flag. we are also not using the adapted data type and projection in `InputResult`. In the third case, we should simply accept the given schema in executeInsert.

Reply:
For my understanding: if the user declares any physical columns, we don't derive anything anymore and use the declared schema as-is? If so, I'd probably rather check for this on the calling side and not call the translator instead of passing a flag to it.

Reply:
I guess we have to take the schema as-is in this case. I wouldn't start with the projection business to exclude columns implicitly. I thought if the flag is set you would simply return the input schema in the last part of the translator. Otherwise you need to have translator logic in `executeInsert`.

Reply:
> Otherwise you need to have translator logic in executeInsert.

Yes, we'd be checking for "does the declared schema contain any physical columns". If it does, we actually end up not doing anything, so an argument could be made that there's no need to call a translator if nothing should be translated. So the question is whether we see the decision of whether to translate this case as the responsibility of the translator.

In general I'm fine with the flag, too. But boolean arguments for control flow make for poor signatures / APIs, especially if it'd already be the fourth argument.

Reply:
I would vote for the flag. Because it simply determines how much the translator should do and when it should stop adjusting the schema.